dace/transformation/estimator/soap/sdg.py

```diff
@@ -0,0 +1,1346 @@
+                        for rng in memlet.subset.ranges
+                            if len(rng[0].free_symbols) > 0])
+                
+                if len(memlet_iter_vars) == 0 and not hasattr(sdg_scope, 'inntermost_ranges'):
+                    return SSA_dim
+                # TODO: experimental. We need have four pools to choose from to find the SSA_dim.
+                # 1. innermost_ranges: these are ranges, either from a map or a loop, that is closest to this statement
+                # 2. loop_ranges
+                # 3. map_ranges
+                # 4. ranges : these are all the ranges in the scope.
+                # It is unclear from which pool we should take it
+                potential_ssa_pool = set(sdg_scope.innermost_ranges) # set(sdg_scope.ranges) - set(sdg_scope.innermost_ranges) 
+                SSA_dim =  [ [dace.symbol(k), dace.symbol(k), 1] \
+                                 for k in potential_ssa_pool - memlet_iter_vars]
+                
+                # we first try to add the SSA_dim as the innermost range (the one from the innermost loop).
+                # If it fails, we then look for the SSA dim in the outer ranges.
+                if not SSA_dim:
+                    SSA_dim =  [ [dace.symbol(k), dace.symbol(k), 1] \
+                            for k in set(sdg_scope.loop_ranges) - memlet_iter_vars]
+        
+        return SSA_dim
+
+    @staticmethod
+    def _memlet_ranges_to_iters(memlet : dace.Memlet) -> dace.Memlet:
+        iter_memlet = copy.deepcopy(memlet)
+        iter_ranges = [(y, y, 1) for (x,y,z) in iter_memlet.subset.ranges]
+        iter_memlet.subset.ranges = iter_ranges
+        return iter_memlet
+            
+
+    
+    # ------------------------------------
+    # helper functions
+    # ------------------------------------
+
+    
+    def plot_SDG(self, filename : str = 'SDG.dot') -> None:
+        nx.nx_pydot.write_dot(self.graph, filename)
+
+
+
+    # ------------------------------------
+    # SDG PARTITIONING
+    # ------------------------------------
+    def perform_loop_swapping(self, node : str, swaplist : Dict[str, str], 
+                              visited: Set[str],
+                              first : bool) -> bool:
+        """
+        Propagates loop variables swapping. E.g., if we have a transient, whose 
+        input edge is [i, j], but the output edge is [i, k], we try to propagate it
+        to all consecutive nodes in the graph, swapping j with k
+        """
+        if node in visited:
+            return True
+        visited |= {node}
+        inv_swaplist = {v : k for k,v in swaplist.items()}
+        
+        if not first:
+            # if not, then the node is NOT a SOAP statement, but some input array
+            if 'st' in self.graph.nodes[node].keys():
+                statement = self.graph.nodes[node]['st']
+                
+                # if '__tmp7' in self.graph.nodes['__tmp8_1']['st'].phis.keys() and \

```

Comment:
Clean-up commented-out code, especially when it is data-dependent (e.g., you have tmp7 and tmp8 here)